mlir/include/mlir/Dialect/MIOpen/XdlopsCodeSelection.h

```diff
@@ -467,221 +442,68 @@ struct XdlopsCodeSelection {
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 32;
-      num_output_blks = 2;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 32;
-      n = 32;
-      k = 2;
-      cycles = 64;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_32x32x4bf16) {
-      group_size = 4;
-      num_groups_blk = 4;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 32;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 32;
-      n = 32;
-      k = 4;
-      cycles = 64;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_16x16x8bf16) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 8;
-      cycles = 32;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_32x32x2f32) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 4;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 2;
-      cycles = 32;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_4x4x2bf16) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 64;
-      num_output_blks = 1;
-      num_regs_xdlops = 4;
-      m = 4;
-      n = 64;
-      k = 2;
-      cycles = 8;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::i32_32x32x8i8) {
-      group_size = 4;
-      num_groups_blk = 4;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 32;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 32;
-      n = 32;
-      k = 8;
-      cycles = 64;
-      k_base = 4;
-    } else if (instr == amdgpu::MFMAInstr::i32_16x16x16i8) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 16;
-      cycles = 32;
-      k_base = 4;
-    } else {
-      llvm_unreachable("Unsupported case as mfmaInstr not selected!\n");
-    }
+    // Rename this field in future refactoring
+    int64_t vectorNumber = imms.size();
+    int64_t mfmasPerXdlops = vectorNumber / (MRepeats * NRepeats);
+    // Since one of MPerXdlops or NPerXdlops can implicitly encode the usage of
+    // broadcasting by the MFMA sequence underlying the xdlops, we take the
+    // smaller of those two values in ordet to determine how many valuse for M
+    // and N are needed for each MFMA.
+    int64_t mfmaNonKDim = std::min(MPerXdlops, NPerXdlops) / mfmasPerXdlops;
```

Comment:
This sounds like a static (non-calculated) value to me. `mfmaNonKDim` sounds like a fundamental value used to derive others, not to be derived from how we define the aggregation of several mfmas.

Reply:
I think this follows a number of other deriving process: To actually check how to set the parameter of a 64 x 64 convolution, we start from a single xdlops instruction and work its way up. However, a lot of the calculation use the opposite direction to take the final product of the derived results and work its way down. If we want to support a new tile size it will make it very hard to come up with those fields of the first big if statement without the inherent values defined in each independent xdlops instruction.